bin/pacignore

```diff
@@ -0,0 +1,238 @@
+  ln="$(grep -n -m 1 '^ *IgnorePkg' "$PACMAN_CONF" | cut -d : -f 1)"
+  if [ -n "$ln" ]; then
+    # Comment all IgnorePkg directives and then
+    # add IgnorePkg with single string at first position
+    sed -i "s/^ *IgnorePkg/#IgnorePkg/g" "$PACMAN_CONF"
+    if ((${#PACIGNORE_IGNOREPKG[@]})); then
+      sed -i "${ln}i IgnorePkg = ${PACIGNORE_IGNOREPKG[*]}" "$PACMAN_CONF"
+    fi
+  else
+    # Add single IgnorePkg directive directly under [options] header
+    sed -i "/^ *\[options\]/a IgnorePkg = ${PACIGNORE_IGNOREPKG[*]}" "$PACMAN_CONF"
+  fi
+}
+
+check_ignorepkg() {
+  # Declare local variable
+  local pkg="$1"
+
+  # Check if package is included in IgnorePkg directive
+  # shellcheck disable=SC2076
+  if [[ " ${PACIGNORE_IGNOREPKG[*]} " =~ " ${pkg} " ]]; then
+    return 0
+  else
+    return 1
+  fi
+}
+
+add_ignorepkg() {
+  # Declare local variable
+  local pkg="$1"
+
+  # Append package to array if it does not exist already
+  if ! check_ignorepkg "$pkg"; then
+    PACIGNORE_IGNOREPKG+=("$pkg")
+  else
+    {
+      eval_gettext "Skipping \$pkg as it is already ignored"
+      printf "\n"
+    } >&2
+    return 1
+  fi
+}
+
+rm_ignorepkg() {
+  # Declare local variables
+  local pkg="$1"
+  local new_array=()
+  local ignorepkg
+
+  # Remove package from array if it exists
+  if check_ignorepkg "$pkg"; then
+    for ignorepkg in "${PACIGNORE_IGNOREPKG[@]}"; do
+      if [ ! "$ignorepkg" == "$pkg" ]; then
+        new_array+=("$ignorepkg")
+      fi
+    done
+
+    # Replace global array with local one
+    PACIGNORE_IGNOREPKG=("${new_array[@]}")
+  else
+    {
+      eval_gettext "Skipping \$pkg as it is was never ignored"
+      printf "\n"
+    } >&2
+    return 1
+  fi
+}
+
+main() {
+  # Declare local variable
+  local pkg exit_code=0 success=0
+  local pacignore_ignorepkg
+
+  # Proceed conditionally given previous exit code
+  if ! pacignore_ignorepkg="$(pacman-conf --config="$PACMAN_CONF" "IgnorePkg" 2>/dev/null)"; then
+    exit_code=1
+    {
+      gettext "Error in reading pacman configuration file"
+      printf "\n"
+    } >&2
+  else
+    # Convert string to array
+    mapfile -t "PACIGNORE_IGNOREPKG" < <(printf "%s" "$pacignore_ignorepkg")
+
+    # Proceed by pacignore subcommands
+    if [ "$PACIGNORE_SUBCOMMAND" == "ls" ]; then
+      printf "%s\n" "${PACIGNORE_IGNOREPKG[@]}"
+    else
+      for pkg; do
+        if ! "${PACIGNORE_SUBCOMMAND}_ignorepkg" "$pkg"; then
+          exit_code=1
+        else
+          success=1
+        fi
+      done
+
+      # Write to pacman configuration file if there is success
+      if ((success)) && [ "$PACIGNORE_SUBCOMMAND" != "check" ]; then
+        write_ignorepkg
+      fi
+    fi
+  fi
+
+  # Return consolidated exit code
+  return "$exit_code"
+}
+
+parse_options() {
+  # Parse the subcommand correctly
+  if [[ "$1" =~ ^(ls|check|add|rm)$ ]]; then
+    PACIGNORE_SUBCOMMAND="$1"
+    shift
+  fi
+
+  # Loop across all CLI arguments and categorize
+  while [[ -n "$1" ]]; do
```

Comment:
I remember I suggested we use Bash's `getopt` instead of rolling our own. This was because, while I'm familiar with (and enjoy) POSIX's `getopts`, I know that it doesn't support long options, but the Bash one does.

I've done some reading since then, and I think it's a mistake. Specifically https://mywiki.wooledge.org/ComplexOptionParsing.

It seems Bash's `getopt` is a mess and no one should use. And the solution for supporting long options is to indeed roll your own. That said, I still prefer POSIX's `getopt` over that, and simply living with single-letter-options-only as a limitation.

The ship has sailed on `downgrade`, IMO, but perhaps as we break out new tools we could go single-letter with POSIX `getopts`?

Reply:
Hmm, I see and did not know about limitations regarding Bash's `getopt` either. I am not against going with POSIX's `getopts` with single-letter options, but I also think we should work on this migration in a separate PR (addressing #102) for all our source code sitting in the `bin` directory.

Would this be fine?

Reply:
>  I also think we should work on this migration in a separate PR

I was not suggesting a migration of `downgrade`s options handling (now or ever), I was suggesting that as we make new things (such as this PR) we make them with that kind of options-handling from the start.

Are you saying you'd prefer to even land this PR as-is and then migrate it later?

Reply:
> I was suggesting that as we make new things (such as this PR) we make them with that kind of options-handling from the start

Ah I see. Hmm, then the question would be whether we start writing the new options-handling now or do we start doing it when we split `downgrade` into the suite of smaller scripts i.e. `downgrade-select` and family.

I see the benefit of doing it now in that we already start this process. I also a downside of this PR becoming too large IMO.

> Are you saying you'd prefer to even land this PR as-is and then migrate it later?

Yes, I would prefer to finish this PR as-is and work on the migration in a separate PR; preferably in the same PR or step when we start splitting `downgrade` into its subordinate scripts. WDYT?

Reply:
> I was not suggesting a migration of downgrades options handling (now or ever)

To this point, I would prefer our CLI option-parsing method to be consistent among our scripts. By this I mean that if we use POSIX `getopts` in `pacignore`, then I think we should be using it in all of our scripts. This is another reason why I am for doing the migration outside of this PR, since it would (in this assumption) involve migration of multiple scripts

Reply:
> I would prefer our CLI option-parsing method to be consistent among our scripts

Aha, that's where we disagree, I think. I don't see any relation between the scripts -- they're just things that call each other, `downgrade` calls `pacignore` no differently than `git` might call `cat` or something like that. Those two don't need any "consistency", right?

This is less true of the way `downgrade` will make use of `downgrade-x` scripts; they have a clear relationship. However, in that case I would look at it as a "high-level" (`downgrade`) vs "low-level" (`downgrade-x`) interface; and I think it'd be reasonable for the high-level to use "nice" (i.e. long flags) options parsing, while the low-level might use something more spartan but easier to implement. So that's one reason I'd be OK with inconsistency between these two.

>  I also see a downside of this PR becoming too large IMO.

But I'm suggesting you do the options parsing only in `pacignore` a simpler way. It would make this PR _smaller_, no? If you can't tell, I'm big on compartmentalization and isolation to reduce cognitive load.

I'm reviewing this PR as establishing a distinct `pacignore` script and not concerning myself with any changes to `downgrade`. In fact, if you were interested in smaller PRs, you could remove everything else and do all the packaging changes and updates to `downgrade` to _use_ `pacignore` in another PR after :thinking: 

All that said, I don't feel strongly on these points and am happy to Approve the PR to let the work evolve. I just wanted to clarify my position to make sure we were on the same page.

Reply:
> So that's one reason I'd be OK with inconsistency between these two.

Thanks for explaining with that reference. I meant consistency in the sense of having a common option parsing standard across all our scripts. But if you frame `downgrade` as a "higher-level" script calling other "lower-level" scripts, then yes I agree it makes sense to have a simpler option parsing framework for the "lower-level" scripts.

Simpler framework aside, another thing holding me back from POSIX's `getopts` was that I have never used it before. Could you perhaps point me to some good starting documentation or suggest a basic diff that could have `getopts` working in `pacignore`?

> It would make this PR smaller, no? 

My bad with a wrong word choice. By "smaller" I meant to minimize the effort in this PR so as to keep things moving and maintain a certain momentum. As mentioned before, I have never used `getopts` so using it in this PR would increase the effort for me to learn and implement it. But this is not an issue since I am on-board for using `getopts`; I just wanted to explain my thought process better.

> If you can't tell, I'm big on compartmentalization and isolation to reduce cognitive load.

Yes

> In fact, if you were interested in smaller PRs, you could remove everything else and do all the packaging changes and updates to downgrade to use pacignore in another PR after thinking

My usual PR workflow is to have enough packed in the PR to allow for a meaningful release afterwards (regardless of whether we actually release or not). This is also why I added the PR checklist with points pertaining to releases. But if you find that this format is too convoluted and increases "cognitive workload", I would be glad to discuss and find an alternative such as having certain PRs for core code changes and other PRs for housekeeping tasks.